text/0000-helpful-error-messages.md

```diff
@@ -0,0 +1,68 @@
+
+The first phase of technical groundwork for improving the error messages was completed in notion-cli/notion#249. The additional work has several phases:
+
+## Unique Identifiers
+
+Update the existing error implementation to support a unique error code assigned to each error type. Additionally, we'll need to update the `Display` implementation to output that error code and the link to the online documentation about that error with the error message.
+
+## Audit Existing Errors
+
+We have a number of existing error messages of varying levels of helpfulness. We should update those to all be at the same level (and potentially split them into multiple errors if necessary).
+
+## Create Error Messages for Unknown Errors
+
+The bulk of the work for this RFC will be in creating new error messages (and codes) and replacing every call to `.unknown()` with a call to `.with_context()` that returns the appropriate error. As of the start of this RFC, there are ~80 uses of `.unknown()`, so this will be a significant undertaking, and likely best handled with a quest issue. For each error, we will need to determine the appropriate context and then craft an error message to explain to the user what went wrong and how best to fix it.
+
+## Online Documentation
+
+We will need to create the online documentation for all of the error messages, and keep it updated as new errors are added. For a day 0 implementation, the documentation could simply be a restatement of the error message, but going forward we should try to add additional context where possible. For example, a relatively common error is `No such file or directory (os error 2)` when trying to execute a command. This very often means that the command being executed couldn't be found in the `PATH`, so we should be able to include that additional information in the online troubleshooting documentation.
+
+# Critique
+[critique]: #critique
+
+The primary point of critique is with the exact voice and style we want to use for the error messages. There are two extremes we could take, though this proposal is for doing a middle-ground that hopefully gets the benefits of each without significant drawbacks.
+
+On the one hand, we could argue that since Notion is a CLI tool, and many of the commands (e.g. `npm run ...`) will be run by automated scripts, we should keep the error messages as concise as possible. This prevents up from filling up logs with human-readable text, while still providing feedback in the form of an error code and short description. The issue with this is that Notion will also be used very often in non-automated contexts. Putting an extra step between the user and useful feedback hurts the user experience and makes it less likely for the helpful information to actually be of use.
```

Comment:
```suggestion
On the one hand, we could argue that since Notion is a CLI tool, and many of the commands (e.g. `npm run ...`) will be run by automated scripts, we should keep the error messages as concise as possible. This prevents us from filling up logs with human-readable text, while still providing feedback in the form of an error code and short description. The issue with this is that Notion will also be used very often in non-automated contexts. Putting an extra step between the user and useful feedback hurts the user experience and makes it less likely for the helpful information to actually be of use.
```